### PR TITLE
Sunset extensions: Open in Editor

### PR DESCRIPTION
**Pull request by vdavid**
Closes https://github.com/sourcegraph/sourcegraph/issues/38378

Added the new settings on the backend, the new UI to the frontend, and a bunch of logic in the background, quite thoroughly unit tested.

Settings are saved on two occasions:
1. When old settings are found and a migration is done
2. When the user presses "Save" on the "Initial setup" popover

Note: Deprecation notices and some thorough testing is missing. I've added these tasks to [the testing doc](https://docs.google.com/document/d/1kAV9zIaXXRLLqqI4DHvmnIX7O9QU6kHPiYddMP-4vck/edit#) to make sure we don't forget these.

## Test plan

- 30+ unit tests
- [5-min Loom](https://www.loom.com/share/5e24a39c033044dabbb20126ba01c0cc) where I demo all the functionalities
  - It was originally meant as a UI demo for Sara, but in it, I went through pretty much all use cases, so I'll link it here too.
  - The docs issue demoed in the Loom is already fixed.

### CODE REVIEW

File: client/web/src/open-in-editor/OpenInEditorPopover.tsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import React from 'react'
+
+import { mdiClose } from '@mdi/js'
+
+import { Button, Icon } from '@sourcegraph/wildcard/out/src'
```
Comment (valerybugakov) [inline client/web/src/open-in-editor/OpenInEditorPopover.tsx:5]:
```suggestion
import { Button, Icon } from '@sourcegraph/wildcard'
```

@vdavid, this change fixes the issue reported [here](https://sourcegraph.slack.com/archives/C01LTKUHRL3/p1662021068019189) for me locally.

File: client/web/src/extensions/components/ActionItemsBar.tsx [hunk 2]
```diff
@@ -223,6 +224,9 @@ export const ActionItemsBar = React.memo<ActionItemsBarProps>(function ActionIte
                         <Icon aria-hidden={true} svgPath={mdiMenuUp} />
                     </Button>
                 )}
+                {extensionsController !== null ? (
```
Comment (philipp-spiess) [inline client/web/src/extensions/components/ActionItemsBar.tsx:227]: This should use the `extensionsAsCoreFeature` flag so that we can turn it on while the legacy extensions are still running.

File: client/web/src/extensions/components/ActionItemsBar.tsx [hunk 3]
```diff
@@ -343,9 +347,9 @@ export const ActionItemsToggle: React.FunctionComponent<React.PropsWithChildren<
                 <div className={classNames(styles.toggleContainer, isOpen && styles.toggleContainerOpen)}>
                     <Tooltip content={`${isOpen ? 'Close' : 'Open'} ${panelName} panel`}>
                         {/**
-                         * This <ButtonLink> must be wrapped with an additional span, since the tooltip currently has an issue that will
-                         * break its onClick handler and it will no longer prevent the default page reload (with no href).
-                         */}
+                          * This <ButtonLink> must be wrapped with an additional span, since the tooltip currently has an issue that will
+                          * break its onClick handler and it will no longer prevent the default page reload (with no href).
+                          */}
```
Comment (philipp-spiess) [inline client/web/src/extensions/components/ActionItemsBar.tsx:352]: prettier will complain about this I reckon :(

File: client/web/src/open-in-editor/OpenInEditorActionItem.tsx [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    platformContext: PlatformContext
+    assetsRoot?: string
+}
+
+export const OpenInEditorActionItem: React.FunctionComponent<OpenInEditorActionItemProps> = props => {
+    const assetsRoot = props.assetsRoot ?? (window.context?.assetsRoot || '')
+
+    const [settingsCascadeOrError, setSettingsCascadeOrError] = useState<SettingsCascadeOrError | undefined>(undefined)
+    const settings = !isErrorLike(settingsCascadeOrError?.final) ? settingsCascadeOrError?.final : undefined
+    const [settingSubscription, setSettingSubscription] = useState<Unsubscribable | null>(null)
```
Comment (philipp-spiess) [inline client/web/src/open-in-editor/OpenInEditorActionItem.tsx:30]: Does this need to be a state? The pattern to initialize it (`useEffect => setState`) is pretty expensive since it requires a full render of the component and will only then start the subscription (in the mean time the app is even responsible already). I wonder if this would be better of with `useObservable => useMemo`?

File: client/web/src/open-in-editor/OpenInEditorActionItem.tsx [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    }, [settingSubscription, props.platformContext.settings, props.platformContext])
+
+    const onClick = useCallback(
+        (event: React.MouseEvent<HTMLElement>) => {
+            event.stopPropagation()
+            if (editor) {
+                openCurrentUrlInEditor(settings?.openInEditor, props.platformContext.sourcegraphURL)
+            } else {
+                togglePopover()
+            }
```
Comment (philipp-spiess) [inline client/web/src/open-in-editor/OpenInEditorActionItem.tsx:80]: Nice this event handler is very clean and understandable.

File: client/web/src/open-in-editor/OpenInEditorActionItem.tsx [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        },
+        [editor, openCurrentUrlInEditor, props.platformContext.sourcegraphURL, settings?.openInEditor, togglePopover]
+    )
+
+    const onSave = useCallback(
+        async (selectedEditorId: EditorId, defaultProjectPath: string): Promise<void> => {
+            const subject = settingsCascadeOrError?.subjects
+                ? settingsCascadeOrError.subjects[settingsCascadeOrError.subjects.length - 1]
```
Comment (philipp-spiess) [inline client/web/src/open-in-editor/OpenInEditorActionItem.tsx:88]: This whole `const subject = ..` line is a bit confusing and we seem to do it many callbacks. Is this something we should prepare in the top level of the component so we can rely on subject being set? Also it's unclear to me what subject refers to here :/ Is it the actual settings object?

File: client/web/src/open-in-editor/OpenInEditorActionItem.tsx [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+                return
+            }
+            await props.platformContext.updateSettings(subject.subject.id, {
+                path: ['openInEditor', 'editorId'],
+                value: selectedEditorId,
+            })
+            await props.platformContext.updateSettings(subject.subject.id, {
+                path: ['openInEditor', 'projectPaths.default'],
+                value: defaultProjectPath,
+            })
```
Comment (philipp-spiess) [inline client/web/src/open-in-editor/OpenInEditorActionItem.tsx:101]: I don't like that these are two updates that happen in sequence. What happens if the browser is closed between the two? 

But I also can't find a nicer way to do that right now so we may just want to think about the case theoretically not that we break something because one option was updated but not the other? 

Edit: So I think we should set the path first and then the ID because if it breaks between those updates, the next time we'd still see that no editor is selected and a use could recover.

File: client/web/src/open-in-editor/build-url.test.ts [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+                    'vscode.useInsiders': true,
+                    'vscode.useSSH': true,
+                    'vscode.remoteHostForSSH': '127.0.0.1',
+                }),
+                baseUrl
+            )
+            expect(url.toString()).toBe(
+                'vscode-insiders://vscode-remote/ssh-remote+127.0.0.1/home/user/projects/sourcegraph/.gitignore:43:0'
+            )
+        })
```
Comment (philipp-spiess) [inline client/web/src/open-in-editor/build-url.test.ts:157]: I have not looked at all test cases but I really like that we have invested in those (especially since you mentioned there are some bugs you found this way?).

File: client/web/src/open-in-editor/build-url.ts [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    range: UIRangeSpec['range'] | undefined,
+    editorSettings: EditorSettings | undefined,
+    sourcegraphBaseUrl: string
+): URL {
+    const editorSettingsErrorMessage = getEditorSettingsErrorMessage(editorSettings, sourcegraphBaseUrl)
+    if (editorSettingsErrorMessage) {
+        throw new TypeError(editorSettingsErrorMessage)
+    }
+
+    const projectPath = getProjectPath(editorSettings || {}) as string // Default is only there to soothe TypeScript
```
Comment (philipp-spiess) [inline client/web/src/open-in-editor/build-url.ts:19]: What does "default" refer to? I would assume it's the initial value but since it's a const it can't change so this doesn't make sense.

File: client/web/src/open-in-editor/build-url.ts [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+        .replace('%line', `${line}`)
+        .replace('%col', `${column}`)
+    return new URL(doReplacements(url, editorSettings?.replacements))
+}
+
+export function getEditorSettingsErrorMessage(
+    editorSettings: EditorSettings | undefined,
+    sourcegraphBaseUrl: string
+): string | undefined {
+    const learnMoreURL = new URL('https://docs.sourcegraph.com/integration/open_in_editor').href
```
Comment (philipp-spiess) [inline client/web/src/open-in-editor/build-url.ts:41]: Isn't this equal to just doing:

```suggestion
    const learnMoreURL = 'https://docs.sourcegraph.com/integration/open_in_editor'
```

File: client/web/src/open-in-editor/build-url.ts [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+                new URL('/user/settings', sourcegraphBaseUrl).href
+            }) to open files. Supported editors: ` + supportedEditors.map(editor => editor.id).join(', ')
+        )
+    }
+    if (editorSettings.editorId === 'custom' && typeof editorSettings['custom.urlPattern'] !== 'string') {
+        return `Add \`custom.urlPattern\` to your user settings for custom editor to open files. [Learn more](${learnMoreURL})`
+    }
+
+    if (editorSettings['vscode.useSSH'] && !editorSettings['vscode.remoteHostForSSH']) {
+        throw new TypeError('`vscode.useSSH` is set to "true" but `vscode.remoteHostForSSH` is not set.')
```
Comment (philipp-spiess) [inline client/web/src/open-in-editor/build-url.ts:77]: What's the conceptual difference here between throwing and returning a string? Should this return a string of the error message instead?

File: client/web/src/open-in-editor/migrate-legacy-settings.test.ts [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+        expect(newSettings.openInEditor).toHaveProperty(['custom.urlPattern'], customUrlPattern)
+        expect(newSettings.openInEditor).toHaveProperty(['vscode.useInsiders'], true)
+        expect(newSettings.openInEditor).toHaveProperty(['vscode.remoteHostForSSH'], vscodeRemoteHost)
+        expect(newSettings.openInEditor).toHaveProperty(['jetbrains.forceApi'], 'builtInServer')
+        expect(newSettings.openInEditor).toHaveProperty(['projectPaths.default'], basePath)
+        expect(newSettings.openInEditor).toHaveProperty(['projectPaths.linux'], linuxBasePath)
+        expect(newSettings.openInEditor).toHaveProperty(['projectPaths.mac'], macBasePath)
+        expect(newSettings.openInEditor).toHaveProperty(['projectPaths.windows'], windowsBasePath)
+        expect(newSettings.openInEditor).toHaveProperty('replacements', replacements)
+    })
```
Comment (philipp-spiess) [inline client/web/src/open-in-editor/migrate-legacy-settings.test.ts:47]: Nice!

File: client/web/src/open-in-editor/migrate-legacy-settings.ts [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+            : null),
+        ...(settings[prefix + '.osPaths.mac'] ? { macBasePath: settings[prefix + '.osPaths.mac'] as string } : null),
+        ...(settings[prefix + '.osPaths.windows']
+            ? { windowsBasePath: settings[prefix + '.osPaths.windows'] as string }
+            : null),
+        ...(settings[prefix + '.replacements']
+            ? { replacements: settings[prefix + '.replacements'] as EditorReplacements }
+            : null),
+    }
+}
```
Comment (philipp-spiess) [inline client/web/src/open-in-editor/migrate-legacy-settings.ts:119]: Oh yeah sure that is exactly what I thought of 😐 🙈

File: client/web/src/open-in-editor/migrate-legacy-settings.ts [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+            delete newSettings[key]
+        }
+    }
+
+    return newSettings
+}
+
+function getOldSettingsInNewFormat(settings: Settings): EditorSettings {
+    const old = {
+        ...readAtomSettings(settings),
```
Comment (philipp-spiess) [inline client/web/src/open-in-editor/migrate-legacy-settings.ts:51]: I feel like creating one curried function for every field is a bit much, should we inline this instead?

e.g.

```suggestion
        ...readLegacySettings(settings, 'openInAtom'),
```

File: client/web/src/open-in-editor/useOpenCurrentUrlInEditor.ts [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+import { buildEditorUrl } from './build-url'
+import { EditorSettings } from './editor-settings'
+
+/**
+ * @returns A function to open the current location in your preferred editor
+ */
+export const useOpenCurrentUrlInEditor = (): ((editorSettings: EditorSettings, sourcegraphURL: string) => void) => {
+    const location = useLocation()
+    const { repoName, filePath, range } = parseBrowserRepoURL(location.pathname)
```
Comment (philipp-spiess) [inline client/web/src/open-in-editor/useOpenCurrentUrlInEditor.ts:15]: I suppose this was a short-cut but I feel like this would be cleaner if we can parse the repo name, file path, and range to the component somehow instead.

It's also currently "buggy" in a sense that it captures the value on the rendering of the component. If the component is not re-rendered when you move to a different document, the values will be stale. We can fix this easily (for now) by moving this line into the callback so that it is evaluated on click and not on render.

File: doc/integration/open_in_editor.md [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+<img src="img/editors/rider.svg" width="32" height="30" alt="Rider logo" />
+<img src="img/editors/rubymine.svg" width="32" height="30" alt="RubyMine logo" />
+<img src="img/editors/webstorm.svg" width="32" height="30" alt="WebStorm logo" />
+<img src="img/editors/sublime.svg" width="32" height="30" alt="Sublime logo" />
+<img src="img/editors/vscode.svg" width="32" height="30" alt="Visual logo" />
+<img src="img/editors/atom.svg" width="32" height="30" alt="Atom logo" />
+
+## Get started
+
+It only needs a minute to set up. If you haven’t used the feature, find this button on the right side of your screen:
```
Comment (philipp-spiess) [inline doc/integration/open_in_editor.md:21]:
```suggestion
If you haven’t used the feature, find this button on the right side of your screen:
```

Let's avoid duplication, this phrase is already used in the paragraph right before

File: doc/integration/open_in_editor.md [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+```json
+{
+  "openInEditor": {
+    "editorId": "vscode",
+    "projectPaths.default": "/Users/USERNAME/Documents/",
+    "replacements": { "sourcegraph-": "" }
+    // vscode://file//Users/USERNAME/Documents/sourcegraph-REST-OF_REPO-NAME/package.json => vscode://file//Users/USERNAME/Documents/REST-OF_REPO-NAME/package.json
+  }
+}
+```
```
Comment (philipp-spiess) [inline doc/integration/open_in_editor.md:226]: Love that you added docs for it. I want to do a more holistic go over the docs for the 4.0 release next week as well, I think you should show this off first as a north-star reference :)

**Review by philipp-spiess — APPROVED** [review]
This is great work!

Two very important things that we need to do before we can land this and start dogfooding:

1. Change the feature flag from `extensionController==null` to use `extensionsAsCoreFeatures`. This is what we have done for the other migrated extensions as well and allows us to show users the newly and improved extensions even if they have to keep using the legacy extension system. (As a side node, the extensions controller will no longer `null` when `enableLegacyExtensions` is true because we'll need it for code intel. In either case, the other flag is better.
2. Make sure to add the open-in-editor extensions (all of them) here:  https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/shared/src/api/client/enabledExtensions.ts?L68-72 This is to ensure that we do not load the old open in extensions if the intersection of a) loading legacy extensions and b) using the migrated extensions is true.

Absolutely hammer PR otherwise:

Schitts Creek Perfect

File: client/web/src/extensions/components/ActionItemsBar.tsx (hunk 3 above)
Comment (vdavid) [reply client/web/src/extensions/components/ActionItemsBar.tsx:352]: Hmm, weird that this happened, I had autoformat enabled for each commit. Thanks for spotting it!

File: client/web/src/open-in-editor/build-url.ts (hunk 9 above)
Comment (vdavid) [reply client/web/src/open-in-editor/build-url.ts:19]: I meant the `|| {}` part. It never happens at this point that `editorSettings` is falsy, so the `|| {}` part is only there to avoid a TypeScript warning. I might just remove the comment because it seems to be more confusing than helpful. WDYT?

Comment (philipp-spiess) [reply client/web/src/open-in-editor/build-url.ts:19]: Ahh. Yeah I agree the comment is a bit confusing to me :/

File: client/web/src/open-in-editor/OpenInEditorActionItem.tsx (hunk 7 above)
Comment (vdavid) [reply client/web/src/open-in-editor/OpenInEditorActionItem.tsx:101]: Yup, I agree. I looked into how we do multiple updates to the settings at once when I implemented this because I'd have preferred to do it at once. But I found that there are only two ways to update settings:
1. To overwrite the whole thing
2. To update one key–value pair
So I went with the second one in this case.
Of course, I could implement a third way, but as we have little time, I just went with this solution, I think it's reliable enough.

In the case when one is updated and not the other, that's of course not great, but _both_ must be set for the editor to be considered as "set up", so in that case, the popover will appear on the next load in either case. So I'd say it doesn't matter much which one we save first.

File: client/web/src/open-in-editor/build-url.test.ts (hunk 8 above)
Comment (vdavid) [reply client/web/src/open-in-editor/build-url.test.ts:157]: Yup, it was very helpful. The logic is not trivial, I think it was not really well thought out for some cases (e.g. handling Windows paths) in the original extensions, and adding these test cases introduced a lot of clarity to what we actually want.

File: client/web/src/open-in-editor/build-url.ts (hunk 11 above)
Comment (vdavid) [reply client/web/src/open-in-editor/build-url.ts:77]: Ah, nice catch, thank you! It used to be all `throw`s, I changed them to `return`s, but somehow missed this one. It shouldn't throw.

File: client/web/src/open-in-editor/migrate-legacy-settings.ts (hunk 13 above)
Comment (vdavid) [reply client/web/src/open-in-editor/migrate-legacy-settings.ts:119]: What did you mean here?

File: doc/integration/open_in_editor.md (hunk 16 above)
Comment (vdavid) [reply doc/integration/open_in_editor.md:21]: Thanks, it was an editing mistake :)

File: client/web/src/open-in-editor/OpenInEditorActionItem.tsx (hunk 7 above)
Comment (philipp-spiess) [reply client/web/src/open-in-editor/OpenInEditorActionItem.tsx:101]: Ah, perfect. That makes sense yep 👏

File: client/web/src/open-in-editor/migrate-legacy-settings.ts (hunk 13 above)
Comment (philipp-spiess) [reply client/web/src/open-in-editor/migrate-legacy-settings.ts:119]: Haha never mind. I tried to be funny but I think my head was full of performance reviews this day and somehow this doesn't make sense at all lol

File: client/web/src/open-in-editor/OpenInEditorActionItem.tsx (hunk 4 above)
Comment (vdavid) [reply client/web/src/open-in-editor/OpenInEditorActionItem.tsx:30]: I've looked into your suggested solution, but it doesn't seem to help. [useObservable](https://sourcegraph.sourcegraph.com/-/editor?remote_url=git%40github.com%3Asourcegraph%2Fsourcegraph.git&branch=dv%2Fsunset-extensions-open-in-editor&file=client%2Fwildcard%2Fsrc%2Fhooks%2FuseObservable.ts&start_row=15&start_col=21&end_row=15&end_col=21&editor=JetBrains&version=v2.0.1) uses `useEffect()` and `useState()` under the hood.

File: client/web/src/open-in-editor/OpenInEditorActionItem.tsx (hunk 6 above)
Comment (vdavid) [reply client/web/src/open-in-editor/OpenInEditorActionItem.tsx:88]: The naming is rather confusing, but it's not my making, I've just followed the terminology in the codebase. We have a few `subject.subject` nonsense elsewhere ([example](https://sourcegraph.sourcegraph.com/-/editor?remote_url=git%40github.com%3Asourcegraph%2Fsourcegraph.git&branch=dv%2Fsunset-extensions-open-in-editor&file=client%2Fshared%2Fsrc%2Fapi%2Fclient%2Fservices%2Fsettings.ts&start_row=41&start_col=25&end_row=41&end_col=25&editor=JetBrains&version=v2.0.1)). Thanks for pointing this out, I've renamed it to "userSettings" because that's what it actually is, and moved it to the top level.

File: client/web/src/open-in-editor/useOpenCurrentUrlInEditor.ts (hunk 15 above)
Comment (vdavid) [reply client/web/src/open-in-editor/useOpenCurrentUrlInEditor.ts:15]: Thanks for catching the bug, yup, you're totally right that it used stale data after navigating between pages.

About the data source: yes, actually, useLocation() was the only place where I found the URL in a format that I could parse in a "standard" way (using "parseBrowserRepoURL" that we use elsewhere). I'd have preferred to grab the repoName, filePath, and range from a pre-parsed location from another component, but I didn't find the data anywhere, so I went for the URl parsing. 🤷‍♂️

**Comment by vdavid** [comment]
@jjinnii 5-min UI Loom demo is here: https://www.loom.com/share/5e24a39c033044dabbb20126ba01c0cc

I've just realized that I've completely missed the "Saving" state (the little spinner on the button) and we're quite in a rush, so I wouldn't like to add it as part of _this_ PR, but please check the Loom and see the behavior for yourself, and if you are on the opinion that we're missing the spinner, I'll add a follow-up "bug fix" issue to add this.

**Comment by vdavid** [comment]
@jjinnii @ryankscott good news! Yesterday in the review meeting I started thinking how much effort it'd actually be to change the data structure to allow for multiple editors to be set, and I figured it was not such a big deal after all. I decided to do this now because if we did dis later, we'd have needed to migrate the current structure to a new one which is extra work.

I've made a [45-sec Loom](https://www.loom.com/share/4642244de56340e381a8a1bf975ff09f) to demo how the new version works.

@jjinnii please confirm if it's fine.

**Comment by jjinnii** [comment]
Thanks @vdavid! It's great that it's no longer a regression 😄 . 

My only comment is that it's not very discoverable, so I'd still like for us to work on the UI of multi-selecting in the actual onboarding flow, as well as we'll have to think of a way to communicate that this is available to users who have already set up this feature immediately after 4.0! 

TL;DR looks great,  I think we still need an issue for the multi select UI. cc: @ryankscott

**Comment by vdavid** [comment]
@jjinnii I've merged this.
Note for myself about the three things undone that I'll need to create issues for:
- Add loading state (low priority)
- Add the ability to select multiple editors from the onboarding flow (high priority)
- Ensure that the popover is aligned exactly to the left of the icon border box (low priority – it's not trivial to do, in theory it should already be aligned)

**Comment by vovakulikov** [comment]
@vdavid I found one file where you used /out/ like imports, was it done intentionally? [build-urls.ts](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Asourcegraph%2Fsourcegraph.git&branch=main&file=client%2Fweb%2Fsrc%2Fopen-in-editor%2Fbuild-url.ts&start_row=2&start_col=24&end_row=2&end_col=24&editor=JetBrains&version=v2.0.1)

**Comment by vdavid** [comment]
@vovakulikov Nice that you spotted this! Nope, not intentional.

On the bright side:
- @philipp-spiess has already prepared a PR where he fixed this
- We discussed how it happened, here: https://github.com/sourcegraph/sourcegraph/pull/41317#discussion_r964725885
- I proposed adding an ESLint rule to avoid this in the future, here: https://sourcegraph.slack.com/archives/C01N83PS4TU/p1662551229868479

**Comment by mrnugget** [comment]
This introduced a regression that causes infinite retries of `updateSettings` calls: https://github.com/sourcegraph/sourcegraph/issues/41522
